File: src/exports/Land.cc
#include "pland/land/Land.h"
#include "pland/PLand.h"
#include "pland/aabb/LandAABB.h"
#include "pland/aabb/LandPos.h"
#include "pland/land/LandContext.h"
#include "pland/utils/JSON.h"

#include "exports/APIHelper.h"

#include "ExportDef.h"

namespace ldapi {


void Export_Class_Land() {
    auto* registry = land::PLand::getInstance().getLandRegistry();

    exportAs("Land_getAABB", [registry](int _landId) -> InternalLandAABB {
        auto land = registry->getLand(_landId);
        if (!land) {
            return {};
        }
        return toLSE<land::LandAABB>(land->getAABB(), land->getDimensionId());
    });

    exportAs("Land_setAABB", [registry](int _landId, InternalLandAABB aabb) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        auto val = toCpp<land::LandAABB>(aabb);
        return land->setAABB(val);
    });

    exportAs("Land_getTeleportPos", [registry](int _landId) -> IntPos {
        auto land = registry->getLand(_landId);
        if (!land) {
            return {};
        }
        return toLSE<land::LandPos>(land->getTeleportPos(), land->getDimensionId());
    });

    exportAs("Land_setTeleportPos", [registry](int _landId, IntPos const& pos) -> void {
        auto land = registry->getLand(_landId);
        if (!land) {
            return;
        }
        land->setTeleportPos(toCpp<land::LandPos>(pos));
    });

    exportAs("Land_getId", [registry](int _landId) -> int {
        auto land = registry->getLand(_landId);
        if (!land) {
            return -1;
        }
        return land->getId();
    });

    exportAs("Land_getDimensionId", [registry](int _landId) -> int {
        auto land = registry->getLand(_landId);
        if (!land) {
            return -1;
        }
        return land->getDimensionId();
    });

    exportAs("Land_getPermTable", [registry](int _landId) -> std::string {
        auto land = registry->getLand(_landId);
        if (!land) {
            return "";
        }
        return toLSE<land::LandPermTable>(land->getPermTable());
    });

    exportAs("Land_setPermTable", [registry](int _landId, std::string const& permTable) -> void {
        auto land = registry->getLand(_landId);
        if (!land) {
            return;
        }
        land->setPermTable(toCpp<land::LandPermTable>(permTable));
    });

    exportAs("Land_getOwner", [registry](int _landId) -> std::string {
        auto land = registry->getLand(_landId);
        if (!land) {
            return "";
        }
        return land->getOwner();
    });

    exportAs("Land_setOwner", [registry](int _landId, std::string const& owner) -> void {
        auto land = registry->getLand(_landId);
        if (!land) {
            return;
        }
        land->setOwner(owner);
    });

    exportAs("Land_getMembers", [registry](int _landId) -> std::vector<std::string> {
        auto land = registry->getLand(_landId);
        if (!land) {
            return {};
        }
        return land->getMembers();
    });

    exportAs("Land_addLandMember", [registry](int _landId, std::string const& member) -> void {
        auto land = registry->getLand(_landId);
        if (!land) {
            return;
        }
        land->addLandMember(member);
    });

    exportAs("Land_removeLandMember", [registry](int _landId, std::string const& member) -> void {
        auto land = registry->getLand(_landId);
        if (!land) {
            return;
        }
        land->removeLandMember(member);
    });

    exportAs("Land_getName", [registry](int _landId) -> std::string {
        auto land = registry->getLand(_landId);
        if (!land) {
            return "";
        }
        return land->getName();
    });

    exportAs("Land_setName", [registry](int _landId, std::string const& name) -> void {
        auto land = registry->getLand(_landId);
        if (!land) {
            return;
        }
        land->setName(name);
    });

    exportAs("Land_getDescribe", [registry](int _landId) -> std::string {
        auto land = registry->getLand(_landId);
        if (!land) {
            return "";
        }
        return land->getDescribe();
    });

    exportAs("Land_setDescribe", [registry](int _landId, std::string const& describe) -> void {
        auto land = registry->getLand(_landId);
        if (!land) {
            return;
        }
        land->setDescribe(describe);
    });

    exportAs("Land_getOriginalBuyPrice", [registry](int _landId) -> int {
        auto land = registry->getLand(_landId);
        if (!land) {
            return 0;
        }
        return land->getOriginalBuyPrice();
    });

    exportAs("Land_setOriginalBuyPrice", [registry](int _landId, int originalBuyPrice) -> void {
        auto land = registry->getLand(_landId);
        if (!land) {
            return;
        }
        land->setOriginalBuyPrice(originalBuyPrice);
    });

    exportAs("Land_is3D", [registry](int _landId) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->is3D();
    });

    exportAs("Land_isOwner", [registry](int _landId, std::string const& uuid) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->isOwner(uuid);
    });

    exportAs("Land_isMember", [registry](int _landId, std::string const& uuid) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->isMember(uuid);
    });

    exportAs("Land_isConvertedLand", [registry](int _landId) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->isConvertedLand();
    });

    exportAs("Land_isOwnerDataIsXUID", [registry](int _landId) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->isOwnerDataIsXUID();
    });

    exportAs("Land_isCollision", [registry](int _landId, IntPos const& pos, int radius) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->isCollision(toCpp<land::LandPos>(pos).as(), radius);
    });

    exportAs("Land_isCollision2", [registry](int _landId, IntPos const& a, IntPos const& b) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->isCollision(toCpp<land::LandPos>(a).as(), toCpp<land::LandPos>(b).as());
    });

    exportAs("Land_isDirty", [registry](int _landId) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->isDirty();
    });

    exportAs("Land_getType", [registry](int _landId) -> int {
        auto land = registry->getLand(_landId);
        if (!land) {
            return -1;
        }
        return static_cast<int>(land->getType());
    });

    exportAs("Land_hasParentLand", [registry](int _landId) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->hasParentLand();
    });

    exportAs("Land_hasSubLand", [registry](int _landId) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->hasSubLand();
    });

    exportAs("Land_isSubLand", [registry](int _landId) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->isSubLand();
    });

    exportAs("Land_isParentLand", [registry](int _landId) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->isParentLand();
    });

    exportAs("Land_isMixLand", [registry](int _landId) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->isMixLand();
    });

    exportAs("Land_isOrdinaryLand", [registry](int _landId) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->isOrdinaryLand();
    });

    exportAs("Land_canCreateSubLand", [registry](int _landId) -> bool {
        auto land = registry->getLand(_landId);
        if (!land) {
            return false;
        }
        return land->canCreateSubLand();
    });

    exportAs("Land_getParentLand", [registry](int _landId) -> int {
        auto land = registry->getLand(_landId);
        if (!land) {
            return -1;
        }
        auto parentLand = land->getParentLand();
        if (!parentLand) {
            return -1;
        }
        return parentLand->getId();
    });

    exportAs("Land_getSubLands", [registry](int _landId) -> std::vector<int> {
        auto land = registry->getLand(_landId);
        if (!land) {
            return {};
        }
        auto             subLands = land->getSubLands();
        std::vector<int> result;
        result.reserve(subLands.size());
        std::transform(subLands.begin(), subLands.end(), std::back_inserter(result), [](auto& subLand) {
            return subLand->getId();
        });
        return result;
    });

    exportAs("Land_getNestedLevel", [registry](int _landId) -> int {
        auto land = registry->getLand(_landId);
        if (!land) {
            return -1;
        }
        return land->getNestedLevel();
    });

    exportAs("Land_getRootLand", [registry](int _landId) -> int {
        auto land = registry->getLand(_landId);
        if (!land) {
            return -1;
        }
        auto rootLand = land->getRootLand();
        if (!rootLand) {
            return -1;
        }
        return rootLand->getId();
    });

    exportAs("Land_getFamilyTree", [registry](int _landId) -> std::vector<int> {
        auto land = registry->getLand(_landId);
        if (!land) {
            return {};
        }
        auto             familyTree = land->getFamilyTree();
        std::vector<int> result;
        result.reserve(familyTree.size());
        std::transform(familyTree.begin(), familyTree.end(), std::back_inserter(result), [](auto& family) {
            return family->getId();
        });
        return result;
    });

    exportAs("Land_getSelfAndAncestors", [registry](int _landId) -> std::vector<int> {
        auto land = registry->getLand(_landId);
        if (!land) {
            return {};
        }
        auto             selfAndAncestors = land->getSelfAndAncestors();
        std::vector<int> result;
        result.reserve(selfAndAncestors.size());
        std::transform(
            selfAndAncestors.begin(),
            selfAndAncestors.end(),
            std::back_inserter(result),
            [](auto& ancestor) { return ancestor->getId(); }
        );
        return result;
    });

    exportAs("Land_getSelfAndDescendants", [registry](int _landId) -> std::vector<int> {
        auto land = registry->getLand(_landId);
        if (!land) {
            return {};
        }
        auto             selfAndDescendants = land->getSelfAndDescendants();
        std::vector<int> result;
        result.reserve(selfAndDescendants.size());
        std::transform(
            selfAndDescendants.begin(),
            selfAndDescendants.end(),
            std::back_inserter(result),
            [](auto& descendant) { return descendant->getId(); }
        );
        return result;
    });

    exportAs("Land_getPermType", [registry](int _landId, std::string const& uuid) -> int {
        auto land = registry->getLand(_landId);
        if (!land) {
            return -1;
        }
        return static_cast<int>(land->getPermType(uuid));
    });
}


} // namespace ldapi

File: ts/imports/Land.ts
import {
    importAs,
    isIntPos,
    LandID,
    LandPermType,
    UUIDs,
} from "../ImportDef.js";
import { LandAABB } from "./LandAABB.js";

export interface LandPermTable {}

export enum LandType {
    Ordinary = 0, // 普通领地(无父、无子)
    Parent = 1, // 父领地(无父、有子)
    Mix = 2, // 混合领地(有父、有子)
    Sub = 3, // 子领地(有父、无子)
}

type InternalLandAABB = [IntPos, IntPos];

export class Land {
    static SYMBOLS = {
        Land_getAABB: importAs("Land_getAABB") as (
            id: LandID
        ) => InternalLandAABB,

        Land_setAABB: importAs("Land_setAABB") as (
            id: LandID,
            aabb: InternalLandAABB
        ) => boolean,

        Land_getTeleportPos: importAs("Land_getTeleportPos") as (
            id: LandID
        ) => IntPos,

        Land_setTeleportPos: importAs("Land_setTeleportPos") as (
            id: LandID,
            pos: IntPos
        ) => void,

        Land_getId: importAs("Land_getId") as (id: LandID) => LandID,

        Land_getDimensionId: importAs("Land_getDimensionId") as (
            id: LandID
        ) => number,

        Land_getPermTable: importAs("Land_getPermTable") as (
            id: LandID
        ) => string,

        Land_setPermTable: importAs("Land_setPermTable") as (
            id: LandID,
            table: string
        ) => void,

        Land_getOwner: importAs("Land_getOwner") as (id: LandID) => UUIDs,

        Land_setOwner: importAs("Land_setOwner") as (
            id: LandID,
            owner: UUIDs
        ) => void,

        Land_getMembers: importAs("Land_getMembers") as (id: LandID) => UUIDs[],

        Land_addLandMember: importAs("Land_addLandMember") as (
            id: LandID,
            uuid: UUIDs
        ) => void,

        Land_removeLandMember: importAs("Land_removeLandMember") as (
            id: LandID,
            uuid: UUIDs
        ) => void,

        Land_getName: importAs("Land_getName") as (id: LandID) => string,

        Land_setName: importAs("Land_setName") as (
            id: LandID,
            name: string
        ) => void,

        Land_getDescribe: importAs("Land_getDescribe") as (
            id: LandID
        ) => string,

        Land_setDescribe: importAs("Land_setDescribe") as (
            id: LandID,
            describe: string
        ) => void,

        Land_getOriginalBuyPrice: importAs("Land_getOriginalBuyPrice") as (
            id: LandID
        ) => number,

        Land_setOriginalBuyPrice: importAs("Land_setOriginalBuyPrice") as (
            id: LandID,
            price: number
        ) => void,

        Land_is3D: importAs("Land_is3D") as (id: LandID) => boolean,
        Land_isOwner: importAs("Land_isOwner") as (
            id: LandID,
            uuid: UUIDs
        ) => boolean,
        Land_isMember: importAs("Land_isMember") as (
            id: LandID,
            uuid: UUIDs
        ) => boolean,

        Land_isConvertedLand: importAs("Land_isConvertedLand") as (
            id: LandID
        ) => boolean,

        Land_isOwnerDataIsXUID: importAs("Land_isOwnerDataIsXUID") as (
            id: LandID
        ) => boolean,

        Land_isCollision: importAs("Land_isCollision") as (
            id: LandID,
            pos: IntPos,
            radius: number
        ) => boolean,

        Land_isCollision2: importAs("Land_isCollision2") as (
            id: LandID,
            pos1: IntPos,
            pos2: IntPos
        ) => boolean,

        Land_isDirty: importAs("Land_isDirty") as (id: LandID) => boolean,
        Land_getType: importAs("Land_getType") as (id: LandID) => number,

        Land_hasParentLand: importAs("Land_hasParentLand") as (
            id: LandID
        ) => boolean,
        Land_hasSubLand: importAs("Land_hasSubLand") as (id: LandID) => boolean,
        Land_isSubLand: importAs("Land_isSubLand") as (id: LandID) => boolean,
        Land_isParentLand: importAs("Land_isParentLand") as (
            id: LandID
        ) => boolean,
        Land_isMixLand: importAs("Land_isMixLand") as (id: LandID) => boolean,
        Land_isOrdinaryLand: importAs("Land_isOrdinaryLand") as (
            id: LandID
        ) => boolean,
        Land_canCreateSubLand: importAs("Land_canCreateSubLand") as (
            id: LandID
        ) => boolean,

        Land_getParentLand: importAs("Land_getParentLand") as (
            id: LandID
        ) => LandID,
        Land_getSubLands: importAs("Land_getSubLands") as (
            id: LandID
        ) => LandID[],
        Land_getNestedLevel: importAs("Land_getNestedLevel") as (
            id: LandID
        ) => number,
        Land_getRootLand: importAs("Land_getRootLand") as (
            id: LandID
        ) => LandID,

        Land_getFamilyTree: importAs("Land_getFamilyTree") as (
            id: LandID
        ) => LandID[],
        Land_getSelfAndAncestors: importAs("Land_getSelfAndAncestors") as (
            id: LandID
        ) => LandID[],
        Land_getSelfAndDescendants: importAs("Land_getSelfAndDescendants") as (
            id: LandID
        ) => LandID[],

        Land_getPermType: importAs("Land_getPermType") as (
            id: LandID,
            uuid: UUIDs
        ) => number,
    };

    readonly mLandId: LandID = -1;
    constructor(id: LandID) {
        this.mLandId = id;
    }

    /**
     * 获取领地AABB范围
     * @returns
     */
    getAABB(): LandAABB | null {
        const result = Land.SYMBOLS.Land_getAABB(this.mLandId);
        if (result.length > 0) {
            return new LandAABB(result[0], result[1]);
        }
        return null;
    }

    /**
     *  修改领地AABB范围（仅限普通领地，且新范围没有任何重叠等冲突）
     * @warning 修改后务必在 LandRegistry 中刷新领地范围
     */
    setAABB(aabb: LandAABB): boolean {
        return Land.SYMBOLS.Land_setAABB(this.mLandId, [aabb.min, aabb.max]);
    }

    getTeleportPos(): IntPos {
        return Land.SYMBOLS.Land_getTeleportPos(this.mLandId);
    }

    setTeleportPos(pos: IntPos): void {
        Land.SYMBOLS.Land_setTeleportPos(this.mLandId, pos);
    }

    getId(): LandID {
        return Land.SYMBOLS.Land_getId(this.mLandId);
    }

    getDimensionId(): number {
        return Land.SYMBOLS.Land_getDimensionId(this.mLandId);
    }

    getPermTable(): LandPermTable | null {
        const result = Land.SYMBOLS.Land_getPermTable(this.mLandId);
        if (result === "") {
            return null;
        }
        return JSON.parse(result);
    }

    setPermTable(table: LandPermTable): void {
        Land.SYMBOLS.Land_setPermTable(this.mLandId, JSON.stringify(table));
    }

    getOwner(): UUIDs {
        return Land.SYMBOLS.Land_getOwner(this.mLandId);
    }

    setOwner(owner: UUIDs): void {
        Land.SYMBOLS.Land_setOwner(this.mLandId, owner);
    }

    getMembers(): UUIDs[] {
        return Land.SYMBOLS.Land_getMembers(this.mLandId);
    }

    addLandMember(uuid: UUIDs): void {
        Land.SYMBOLS.Land_addLandMember(this.mLandId, uuid);
    }

    removeLandMember(uuid: UUIDs): void {
        Land.SYMBOLS.Land_removeLandMember(this.mLandId, uuid);
    }

    getName(): string {
        return Land.SYMBOLS.Land_getName(this.mLandId);
    }

    setName(name: string): void {
        Land.SYMBOLS.Land_setName(this.mLandId, name);
    }

    getDescribe(): string {
        return Land.SYMBOLS.Land_getDescribe(this.mLandId);
    }

    setDescribe(describe: string): void {
        Land.SYMBOLS.Land_setDescribe(this.mLandId, describe);
    }

    getOriginalBuyPrice(): number {
        return Land.SYMBOLS.Land_getOriginalBuyPrice(this.mLandId);
    }

    setOriginalBuyPrice(price: number): void {
        Land.SYMBOLS.Land_setOriginalBuyPrice(this.mLandId, price);
    }

    is3D(): boolean {
        return Land.SYMBOLS.Land_is3D(this.mLandId);
    }

    isOwner(uuid: UUIDs): boolean {
        return Land.SYMBOLS.Land_isOwner(this.mLandId, uuid);
    }

    isMember(uuid: UUIDs): boolean {
        return Land.SYMBOLS.Land_isMember(this.mLandId, uuid);
    }

    isConvertedLand(): boolean {
        return Land.SYMBOLS.Land_isConvertedLand(this.mLandId);
    }

    isOwnerDataIsXUID(): boolean {
        return Land.SYMBOLS.Land_isOwnerDataIsXUID(this.mLandId);
    }

    isCollision(pos: IntPos, radius: number): boolean;
    isCollision(pos1: IntPos, pos2: IntPos): boolean;
    isCollision(pos1: IntPos, pos2: any): boolean {
        if (typeof pos2 === "number") {
            return Land.SYMBOLS.Land_isCollision(this.mLandId, pos1, pos2);
        } else if (isIntPos(pos2)) {
            return Land.SYMBOLS.Land_isCollision2(this.mLandId, pos1, pos2);
        }
        throw new TypeError("pos2 must be IntPos or number");
    }

    isDirty(): boolean {
        return Land.SYMBOLS.Land_isDirty(this.mLandId);
    }

    getType(): LandType | null {
        const result = Land.SYMBOLS.Land_getType(this.mLandId);
        if (result === -1) {
            return null;
        }
        return result as LandType;
    }

    hasParentLand(): boolean {
        return Land.SYMBOLS.Land_hasParentLand(this.mLandId);
    }

    hasSubLand(): boolean {
        return Land.SYMBOLS.Land_hasSubLand(this.mLandId);
    }

    isSubLand(): boolean {
        return Land.SYMBOLS.Land_isSubLand(this.mLandId);
    }

    isParentLand(): boolean {
        return Land.SYMBOLS.Land_isParentLand(this.mLandId);
    }

    isMixLand(): boolean {
        return Land.SYMBOLS.Land_isMixLand(this.mLandId);
    }

    isOrdinaryLand(): boolean {
        return Land.SYMBOLS.Land_isOrdinaryLand(this.mLandId);
    }

    canCreateSubLand(): boolean {
        return Land.SYMBOLS.Land_canCreateSubLand(this.mLandId);
    }

    getParentLand(): Land | null {
        const result = Land.SYMBOLS.Land_getParentLand(this.mLandId);
        if (result === -1) {
            return null;
        }
        return new Land(result);
    }

    getSubLands(): Land[] {
        return Land.SYMBOLS.Land_getSubLands(this.mLandId).map(
            (id) => new Land(id)
        );
    }

    getNestedLevel(): number {
        return Land.SYMBOLS.Land_getNestedLevel(this.mLandId);
    }

    getRootLand(): Land | null {
        const result = Land.SYMBOLS.Land_getRootLand(this.mLandId);
        if (result === -1) {
            return null;
        }
        return new Land(result);
    }

    getFamilyTree(): Land[] {
        return Land.SYMBOLS.Land_getFamilyTree(this.mLandId).map(
            (id) => new Land(id)
        );
    }

    getSelfAndAncestors(): Land[] {
        return Land.SYMBOLS.Land_getSelfAndAncestors(this.mLandId).map(
            (id) => new Land(id)
        );
    }

    getSelfAndDescendants(): Land[] {
        return Land.SYMBOLS.Land_getSelfAndDescendants(this.mLandId).map(
            (id) => new Land(id)
        );
    }

    getPermType(uuid: UUIDs): LandPermType {
        return Land.SYMBOLS.Land_getPermType(this.mLandId, uuid);
    }
}

Object.freeze(Land.SYMBOLS);
